test/hotspot/jtreg/gc/TestNativeReferenceGet.java

```diff
@@ -0,0 +1,189 @@
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.List;
+import jdk.internal.access.JavaLangRefAccess;
+import jdk.internal.access.SharedSecrets;
+import jdk.test.whitebox.WhiteBox;
+
+public final class TestNativeReferenceGet {
+
+    private static final WhiteBox WB = WhiteBox.getWhiteBox();
+
+    private static void gcUntilOld(Object o) {
+        while (!WB.isObjectInOldGen(o)) {
+            WB.fullGC();
+        }
+    }
+
+    private static final class TestObject {
+        public final int value;
+
+        public TestObject(int value) {
+            this.value = value;
+        }
+    }
+
+    private static final ReferenceQueue<TestObject> queue =
+        new ReferenceQueue<TestObject>();
+
+    private static final class Ref extends WeakReference<TestObject> {
+        public Ref(TestObject obj) {
+            super(obj, queue);
+        }
+    }
+
+    private static final int NUM_REFS = 100;
+
+    private static List<Ref> references = null;
+    private static List<TestObject> referents = null;
+
+    // Create all the objects used by the test, and ensure they are all in the
+    // old generation.
+    private static void setup() {
+        references = new ArrayList<Ref>(NUM_REFS);
+        referents = new ArrayList<TestObject>(NUM_REFS);
+
+        for (int i = 0; i < NUM_REFS; ++i) {
+            TestObject obj = new TestObject(i);
+            referents.add(obj);
+            references.add(new Ref(obj));
+        }
+
+        gcUntilOld(references);
+        gcUntilOld(referents);
+        for (int i = 0; i < NUM_REFS; ++i) {
+            gcUntilOld(references.get(i));
+            gcUntilOld(referents.get(i));
+        }
+    }
+
+    // Discard all the strong references.
+    private static void clearReferents() {
+        // Not using List.clear() because it doesn't document null'ing elements.
+        for (int i = 0; i < NUM_REFS; ++i) {
+            referents.set(i, null);
+        }
+    }
+
+    // Create new strong references from the weak references, by using the
+    // native method implementation of Reference.get() and recording the value
+    // in references.
+    private static void strengthenReferents() {
+        for (int i = 0; i < NUM_REFS; ++i) {
+            referents.set(i, references.get(i).get());
+        }
+    }
+
+    private static void checkQueue() {
+        if (queue.poll() != null) {
+            throw new RuntimeException("Reference enqueued");
+        }
+    }
+
+    private static void check() {
+        // None of the references should have been cleared and enqueued,
+        // because we have strong references to all the referents.
+        try {
+            JavaLangRefAccess jlra = SharedSecrets.getJavaLangRefAccess();
+            while (jlra.waitForReferenceProcessing()) {
+                checkQueue();
+            }
+            checkQueue();       // One last check after refproc complete.
+        } catch (InterruptedException e) {
```

Comment:
Rather than using Reference.remove with a timeout, I've changed this to use waitForReferenceProcessing.
That removes false passes (from reference processing being slow to deliver) and also removes the delay
until timeout for the passing case.

Reply:
https://github.com/openjdk/jdk/pull/24527 is proposing a different way to access
`waitForReferenceProcessing`.  I'll wait for that.

Reply:
https://github.com/openjdk/jdk/pull/24527 is proposing a different way to access
`waitForReferenceProcessing`.  I'll wait for that.

Reply:
Test has now been changed to use the new `WhiteBox::waitForReferenceProcessing()`.
So how about some reviews...